packages/grafana-ui/src/components/Text/Text.tsx

```diff
@@ -0,0 +1,130 @@
+  /** What typograpy variant should be used for the component. Only use if default variant for the defined 'as' is not what is needed */
+  variant?: keyof ThemeTypographyVariantTypes;
+  /** Override the default weight for the used variant */
+  weight?: 'light' | 'regular' | 'medium' | 'bold';
+  /** Color to use for text */
+  color?: keyof GrafanaTheme2['colors']['text'] | 'error' | 'success' | 'warning' | 'info';
+  /** Use to cut the text off with ellipsis if there isn't space to show all of it. On hover shows the rest of the text */
+  truncate?: boolean;
+  /** Whether to align the text to left, center or right */
+  textAlignment?: CSSProperties['textAlign'];
+  /** Margin to set on the component. Remember, it does not work with inline elements such as 'span'. */
+  margin?: string | number;
+  /** Class to apply to the component */
+  className?: string;
+  children: React.ReactNode;
+}
+
+export const Text = ({
+  as = 'span',
+  variant,
+  weight,
+  color,
+  truncate,
+  textAlignment,
+  margin,
+  className,
+  children,
+}: TextProps) => {
+  const styles = useStyles2(
+    useCallback(
+      (theme) => {
+        let styleVariant: keyof ThemeTypographyVariantTypes;
+        if (!variant) {
+          if (as === 'span' || as === 'legend') {
+            styleVariant = 'bodySmall';
+          } else if (as === 'p') {
+            styleVariant = 'body';
+          } else {
+            styleVariant = as;
+          }
+        } else {
+          styleVariant = variant;
+        }
+        return getTextStyles(theme, styleVariant, color, weight, truncate, textAlignment, margin);
+      },
+      [color, margin, textAlignment, truncate, weight, as, variant]
+    )
+  );
+
+  return createElement(
+    as,
+    {
+      className: cx(className, styles),
+    },
+    children
+  );
+};
+
+Text.displayName = 'Text';
+
+const getTextStyles = (
+  theme: GrafanaTheme2,
+  variant: keyof ThemeTypographyVariantTypes,
+  color?: TextProps['color'],
+  weight?: TextProps['weight'],
+  truncate?: TextProps['truncate'],
+  textAlignment?: TextProps['textAlignment'],
+  margin?: TextProps['margin']
+) => {
+  return css([
+    {
+      ...theme.typography[variant],
+    },
+    color && {
+      color: customColor(color, theme),
+    },
+    weight && {
+      fontWeight: customWeight(weight, theme),
+    },
+    truncate && {
+      overflow: 'hidden',
+      textOverflow: 'ellipsis',
+      whiteSpace: 'nowrap',
+    },
+    margin && {
+      margin,
+    },
+    textAlignment && {
+      textAlign: textAlignment,
+    },
+  ]);
+};
+
+const customWeight = (weight: TextProps['weight'], theme: GrafanaTheme2) => {
+  switch (weight) {
+    case 'bold':
+      return theme.typography.fontWeightBold;
+    case 'medium':
+      return theme.typography.fontWeightMedium;
+    case 'light':
+      return theme.typography.fontWeightLight;
+    default:
+      return theme.typography.fontWeightRegular;
+  }
+};
```

Comment:
> We should make this an 'exhaustive switch' to make typescript do the work of checking that we didn't forget anything. by specifying the return type of the function, and explicitly including regular and undefined cases (rather than relying on default)
> 
> See [this playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgApQPYAcDOyDeAsAFDJnIDuEwA5gBZgD8AXMgEQBGGANgCZvIAPuwC2EXsACuIgcLZQINSdzhRZ7brQZsA3CQC+JEjEkgEYYBhDIEknGAwiA6tXpgAFFS1hW6bDgBtNi83NgBdAEpWEGkOaAISMhwKYDAEOmRPVwYIhNJyBDgcFAByLj4S5kTyMgUwSShrAA4ABha9fLJC4uQSsQlpSuqauobrAFY2jpru0s03Ic7yUcbkACYp4Zsi0oUlFShFmZ3kU14IGFBxKqXaiHrVgBZN-MNifQ6SAHov5AB5fzIACSyBgGCgNAwYGQDmQoAQ3Ek52QcEo2WhYKgyAAMt4AITGUzmSzWWz2RwubwAIUkYCBADFwZCwABlRz3OigGhZby+TC4IIhbSRaKxeJEfLJVLpTJCsC5CXHHplHi8I41ZArZovJWlfpSETqkb3MbISbtLazXp7ZSqI3bHpnC5XXg3DWak1PHXIN4fIA)

Why are you using `400` instead of trusting the theme using `theme.typography.fontWeightRegular`?  If we use a number rather than this variable and someday we change the weight in the theme, we have to be aware of updating this function instead of it being updated automatically 🤔

Reply:
(sorry, don't pay too much attention to _that_ part of the playground - that was just to get a working demo that shows an example of exhaustive switch check)

Reply:
We would probably do something like:

```ts
const customWeight = (weight: TextProps['weight'], theme: GrafanaTheme2): number => {
  switch (weight) {
    case 'bold':
      return theme.typography.fontWeightBold;
    case 'medium':
      return theme.typography.fontWeightMedium;
    case 'light':
      return theme.typography.fontWeightLight;
    case 'regular':
    case undefined:
      return theme.typography.fontWeightRegular;
  }
};
```

Reply:
Ah, ok, I thought that had a reason 🤔 
Cool, that's what is on the code after applying your suggestion (the part that I agreed with 😉 )